python/llm/src/ipex_llm/transformers/models/llama.py

```diff
@@ -1318,33 +1328,65 @@ def llama_attention_forward_4_36_original(
 
 
 def native_sdp(query, key, value, attention_mask,
-               bsz, q_len, kv_seq_len, head_dim, num_heads):
-    attn_weights = torch.matmul(query.to(key.dtype),
-                                key.transpose(2, 3)) / math.sqrt(head_dim)
+               bsz, q_len, kv_seq_len, head_dim, num_heads, output_attentions):
+    if should_split_qkv_tensor(query, output_attentions):
+        return native_sdp_split_qkv_tensor(query, key, value, attention_mask,
+                                           bsz, q_len, kv_seq_len, head_dim)
+    else:
+        attn_weights = torch.matmul(query.to(key.dtype),
+                                    key.transpose(2, 3)) / math.sqrt(head_dim)
 
-    attn_weights_size = (bsz, num_heads, q_len, kv_seq_len)
-    if attn_weights.size() != attn_weights_size:
-        invalidInputError(False,
-                          f"Attention weights should be of size {attn_weights_size}, "
-                          f"but is {attn_weights.size()}")
+        attn_weights_size = (bsz, num_heads, q_len, kv_seq_len)
+        if attn_weights.size() != attn_weights_size:
+            invalidInputError(False,
+                              f"Attention weights should be of size {attn_weights_size}, "
+                              f"but is {attn_weights.size()}")
+
+        if attention_mask is not None:
+            attn_mask_size = (bsz, 1, q_len, kv_seq_len)
+            if attention_mask.size() != attn_mask_size:
+                invalidInputError(False,
+                                  f"Attention mask should be of size {attn_mask_size}, "
+                                  f"but is {attention_mask.size()}")
+            attn_weights = attn_weights + attention_mask
 
-    if attention_mask is not None:
-        attn_mask_size = (bsz, 1, q_len, kv_seq_len)
-        if attention_mask.size() != attn_mask_size:
+        if kv_seq_len >= 2048:
+            # for memory considerations, do not upcast attention to fp32 for long sequences
+            attn_weights = nn.functional.softmax(attn_weights, dim=-1)
+        else:
+            # upcast attention to fp32
+            attn_weights = nn.functional.softmax(attn_weights, dim=-1,
+                                                 dtype=torch.float32).to(value.dtype)
+        attn_output = torch.matmul(attn_weights, value)
+        return attn_output, attn_weights
+
+
+def native_sdp_split_qkv_tensor(query, key, value, attention_mask,
+                                bsz, q_len, kv_seq_len, head_dim):
+    query_sp = torch.split(query.to(key.dtype), 16, dim=1)
```

Comment:
does sp mean split? Can we change for a better name to understand?

Reply:
fixed.